src/gt4py/next/type_system_2/traits.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import annotations
+
+import abc
+from typing import Any, Optional
+from collections.abc import Sequence
+
+import dataclasses
+
+
+class Trait:
+    ...
+
+
+class SignednessTrait(Trait):
+    @abc.abstractmethod
+    def is_signed(self) -> bool:
+        ...
+
+
+class FromTrait(Trait):
+    @abc.abstractmethod
+    def is_constructible_from(self, ty: Any) -> bool:
+        ...
+
+
+class ToTrait(Trait):
+    @abc.abstractmethod
+    def is_convertible_to(self, ty: Any) -> bool:
+        ...
```

Comment:
Note to myself: check why both directions exist? what happens if they contradict?

Reply:
It's so that the conversion A -> B can be defined in the code of either A or B, but normally not both so that you avoid the contradictions. You'll choose the implementation site depending on whether modA imports modB, or the other way around. If modA imports modB, then you'd implement `A::is_convertible_to(B)`, otherwise you would implement `B::is_constructible_from(A)`.